quickwit-cluster/docs/kubernetes-discovery.md

```diff
@@ -0,0 +1,220 @@
+  - with a label selector matching both the `$release-indexing` and `$release-search` pods, as a unified cluster
+
+## Proposed solution
+
+The following action items are ordered by priority, and could be independently implemented in several releases.
+Alternatives are discussed in the next section, with a pros/cons comparison.
+
+### 1a. Source the sibling pods as seeds via DNS enumeration on startup
+
+A new `peer_seed_service` configuration option is added to Quickwit. It is exclusive to the existing `peer_seeds`,
+and setting both will result on a validation error. `quickwit-cluster::start_cluster_service` is updated to check
+this option. If it is not empty:
+
+- a DNS enumeration request is issued, that will return the IP addresses of all the live pods in the cluster,
+- if the DNS request fails or if the result is empty, Quickwit fails to start. Kubernetes will restart the
+  container, allowing to retry the request when more pods are created. In case of repeated failures (misconfiguration,
+  or DNS error), the CrashLoopBackoff mechanism will engage, reducing the pressure on the DNS server by keeping
+  pods down longer between retries,
+- if not empty, the response is used to populate the `seed_nodes` argument passed to `scuttlebutt`.
+
+Due to the eventually consistent nature of the k8s control plane, there will be a delay before all pods are exposed
+in the DNS record. We do have guarantees that the responses to the different pods will be consistent. Pods will
+restart until one IP is registered, and the same first peer will be returned to all pods, avoiding a split brain.
+
+When adding nodes to the cluster (by scaling up the deployment/statefulset), new nodes will retrieve
+[up to 1000](https://kubernetes.io/docs/concepts/services-networking/service/#over-capacity-endpoints)
+IPs as seeds by querying the service. Existing nodes will not update their seeds, but discover the new nodes
+via the gossip messages.
+
+If an existing node restarts (due to an out-of-memory error for example), it will request up-to-date seeds again
+from the DNS servers, discarding its previous state. This is better than running the DNS enumeration in an
+[init container](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/), which would not
+update the seeds on container restart.
+
+### 2. Gate cluster membership on new a `cluster_name` property
+
+Redis, Cassandra and Elasticsearch all had to solve the IP reuse problem before. They all settled on the simplest and
+most resilient solution: adding a `cluster_name` property to the nodes, and forbidding nodes to join if their
+advertised cluster name does not match the other peers. This mechanism can be implemented in `scuttlebutt` with
+the following changes:
+
+- adding a new required argument `cluster_name` to the `ScuttleServer::spawn` function (String type)
+- copying that value as a new  `cluster_name` field to the `Digest` object
```

Comment:
Syn or Digest?

Reply:
`Syn` could be enough, but implementing the check at the `Digest` level also allows nodes to avoid being "poisoned" by a bad `SynAck` message from a buggy release.

Reply:
(Receiving an Ack message is what triggers a write.)
Let s simply put it on all message then maybe?

Reply:
SGTM